### PR TITLE
TSK-970

### PR DESCRIPTION
**Pull request by gitgoodjhe**
-Added feature to retrieve process variables from a camunda task and write them into the payload of the taskana-outbox

-TaskanaAdapter now adds a localVariable when completing a task in camunda called "completedByTaskanaAdapter"

-Adjusted TaskanaTaskListener to check for local variable "completedByTaskanaAdapter" and only inserting "complete" event, if it is not present

-Added process models & tests for complex process variables

-Deleted CamundaVariableRetriever as it is not needed anymore

-Adjusted Classes that used the CamundaVariableRetriever

### CODE REVIEW

File: taskana-adapter-camunda-tasklistener/src/main/java/pro/taskana/adapter/camunda/tasklistener/TaskanaTaskListener.java [hunk 1]
```diff
@@ -22,6 +23,9 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import pro.taskana.adapter.camunda.dto.VariableValueDto;
+import pro.taskana.adapter.camunda.mapper.JacksonConfigurator;
+
 public class TaskanaTaskListener implements TaskListener {
```
Comment (BerndBreier) [inline taskana-adapter-camunda-tasklistener/src/main/java/pro/taskana/adapter/camunda/tasklistener/TaskanaTaskListener.java]: Please add Javadoc on class level

File: taskana-adapter-test/src/test/java/pro/taskana/adapter/integration/TestTaskAcquisition.java [hunk 2]
```diff
@@ -83,40 +89,73 @@ public void multiple_user_task_process_instances_started_in_camunda_via_rest_sho
...
         String variables = "\"variables\": {\"amount\": {\"value\":555, \"type\":\"long\"},\"item\": {\"value\": \"item-xyz\"}}";
         String processInstanceId = this.camundaProcessengineRequester
-            .startCamundaProcessAndReturnId("simple_user_task_process", variables);
+                .startCamundaProcessAndReturnId("simple_user_task_process", variables);
         List<String> camundaTaskIds = this.camundaProcessengineRequester
-            .getTaskIdsFromProcessInstanceId(processInstanceId);
+                .getTaskIdsFromProcessInstanceId(processInstanceId);
+
+        Thread.sleep((long) (this.adapterTaskPollingInterval * 1.2));
+
```
Comment (BerndBreier) [inline taskana-adapter-test/src/test/java/pro/taskana/adapter/integration/TestTaskAcquisition.java]: The thread should sleep 1.2 times the polling interval

File: taskana-adapter-test/src/test/java/pro/taskana/adapter/integration/TestTaskAcquisition.java [hunk 3]
```diff
@@ -83,40 +89,73 @@ public void multiple_user_task_process_instances_started_in_camunda_via_rest_sho
...
-            // rather use SameJSONAs.sameJSONAs from hamcrest-json to compare Json strings independent of child order
-            Assert.assertThat(
-                assumedVariablesString,
-                SameJSONAs.sameJSONAs(taskanaVariablesString));
-        }
+    @WithAccessId(
+            userName = "teamlead_1",
+            groupNames = {"admin"})
+    @Test
+    public void task_with_complex_variables_from_parent_execution_should_result_in_taskanaTasks_with_those_variables_in_custom_attributes()
```
Comment (BerndBreier) [inline taskana-adapter-test/src/test/java/pro/taskana/adapter/integration/TestTaskAcquisition.java]: Please add a testcase where a variable contains a String that contains characters that must be escaped in Json like
 
\b  Backspace (ascii code 08)
\f  Form feed (ascii code 0C)
\n  New line
\r  Carriage return
\t  Tab
\"  Double quote
\\  Backslash character
Just to be sure that we transfer these variables correctly to taskana ...

File: taskana-adapter-test/src/test/java/pro/taskana/adapter/integration/TestTaskAcquisition.java [hunk 4]
```diff
@@ -140,4 +180,35 @@ public void process_instance_with_multiple_executions_should_result_in_multiple_
             assertEquals(taskanaTaskExternalId, camundaTaskId);
         }
     }
-}
\ No newline at end of file
+
+    private void createTaskanaTaskAndVerifyTaskVariables(String camundaTaskId, String assumedVariablesString) {
+
```
Comment (BerndBreier) [inline taskana-adapter-test/src/test/java/pro/taskana/adapter/integration/TestTaskAcquisition.java]: please rename this method: the testcase doesn't create a taskana task, it rather relies on the adapter to have already created a taskana task. The testcase simply checks that the taskana task exists and has the correct content ....

File: taskana-adapter-test/src/test/java/pro/taskana/impl/ProcessVariableTestObjectTwo.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
+package pro.taskana.impl;
+
+import java.io.Serializable;
+import java.util.Date;
+
+public class ProcessVariableTestObjectTwo implements Serializable {
```
Comment (BerndBreier) [inline taskana-adapter-test/src/test/java/pro/taskana/impl/ProcessVariableTestObjectTwo.java]: Please add Javadoc

File: taskana-adapter-test/src/test/java/pro/taskana/impl/ProcessVariableTestObject.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
+package pro.taskana.impl;
+
+import java.io.Serializable;
+
+public class ProcessVariableTestObject implements Serializable {
```
Comment (BerndBreier) [inline taskana-adapter-test/src/test/java/pro/taskana/impl/ProcessVariableTestObject.java]: Please add Javadoc

File: taskana-adapter-test/src/test/java/pro/taskana/impl/ComplexProcessVariableSetter.java [hunk 7]
```diff
@@ -0,0 +1,25 @@
+package pro.taskana.impl;
+
+import java.text.SimpleDateFormat;
+import java.util.Date;
+
+import org.camunda.bpm.engine.delegate.DelegateExecution;
+import org.camunda.bpm.engine.delegate.JavaDelegate;
+
+public class ComplexProcessVariableSetter implements JavaDelegate {
```
Comment (BerndBreier) [inline taskana-adapter-test/src/test/java/pro/taskana/impl/ComplexProcessVariableSetter.java]: Please add Javadoc

File: taskana-adapter-camunda-tasklistener/src/main/java/pro/taskana/adapter/camunda/dto/VariableValueDto.java [hunk 8]
```diff
@@ -0,0 +1,44 @@
+package pro.taskana.adapter.camunda.dto;
+
+import com.fasterxml.jackson.annotation.JsonRawValue;
+
+import java.util.Map;
+
+public class VariableValueDto {
```
[the remark this answers is not in the document]
Comment (gitgoodjhe) [reply taskana-adapter-camunda-tasklistener/src/main/java/pro/taskana/adapter/camunda/dto/VariableValueDto.java]: Done

File: taskana-adapter-camunda-tasklistener/src/main/java/pro/taskana/adapter/camunda/mapper/JacksonConfigurator.java [hunk 9]
```diff
@@ -0,0 +1,22 @@
+package pro.taskana.adapter.camunda.mapper;
+
+import java.text.SimpleDateFormat;
+
+import com.fasterxml.jackson.databind.DeserializationFeature;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.SerializationFeature;
+
+public class JacksonConfigurator {
```
[the remark this answers is not in the document]
Comment (gitgoodjhe) [reply taskana-adapter-camunda-tasklistener/src/main/java/pro/taskana/adapter/camunda/mapper/JacksonConfigurator.java]: Done

File: taskana-adapter-camunda-tasklistener/src/main/java/pro/taskana/adapter/camunda/tasklistener/TaskanaTaskListener.java [hunk 10]
```diff
@@ -185,11 +200,77 @@ private String getReferencedTaskJson(DelegateTask delegateTask) {
...
+                LOGGER.warn("Caught {} while trying to create JSON-String out of process variable object", ex);
+            }
+        }
+    }
+
+    private List<String> splitProcessVariablesString(String processVariablesConcatenated) {
+        List<String> processVariables = Arrays.asList(processVariablesConcatenated.trim().split("\\s*,\\s*"));
+        return processVariables;
+    }
+
```
[the remark this answers is not in the document]
Comment (gitgoodjhe) [reply taskana-adapter-camunda-tasklistener/src/main/java/pro/taskana/adapter/camunda/tasklistener/TaskanaTaskListener.java]: Done

File: taskana-adapter-camunda-tasklistener/src/main/java/pro/taskana/adapter/camunda/tasklistener/TaskanaTaskListener.java (hunk 1 above)
Comment (gitgoodjhe) [reply taskana-adapter-camunda-tasklistener/src/main/java/pro/taskana/adapter/camunda/tasklistener/TaskanaTaskListener.java]: Done

File: taskana-adapter-test/src/test/java/pro/taskana/adapter/integration/TestTaskAcquisition.java (hunk 2 above)
Comment (gitgoodjhe) [reply taskana-adapter-test/src/test/java/pro/taskana/adapter/integration/TestTaskAcquisition.java]: Done

File: taskana-adapter-test/src/test/java/pro/taskana/adapter/integration/TestTaskAcquisition.java (hunk 3 above)
Comment (gitgoodjhe) [reply taskana-adapter-test/src/test/java/pro/taskana/adapter/integration/TestTaskAcquisition.java]: Done

File: taskana-adapter-test/src/test/java/pro/taskana/impl/ProcessVariableTestObjectTwo.java (hunk 5 above)
Comment (gitgoodjhe) [reply taskana-adapter-test/src/test/java/pro/taskana/impl/ProcessVariableTestObjectTwo.java]: Done

File: taskana-adapter-test/src/test/java/pro/taskana/adapter/integration/TestTaskAcquisition.java (hunk 4 above)
Comment (gitgoodjhe) [reply taskana-adapter-test/src/test/java/pro/taskana/adapter/integration/TestTaskAcquisition.java]: Done

File: taskana-adapter-test/src/test/java/pro/taskana/impl/ComplexProcessVariableSetter.java (hunk 7 above)
Comment (gitgoodjhe) [reply taskana-adapter-test/src/test/java/pro/taskana/impl/ComplexProcessVariableSetter.java]: Done

File: taskana-adapter-camunda-tasklistener/src/main/java/pro/taskana/adapter/camunda/tasklistener/TaskanaTaskListener.java [hunk 11]
```diff
@@ -160,92 +170,63 @@ private void prepareAndExecuteStatement(Connection connection, DelegateTask dele
...
-        }
-
-        referencedTaskJsonBuilder.append("\",\"variables\":\"{")
-            .append(getProcessVariables(delegateTask));
+    private String getReferencedTaskJson(DelegateTask delegateTask) throws JsonProcessingException {
 
-        referencedTaskJsonBuilder.append("}");
-
-        return referencedTaskJsonBuilder.toString();
+        ObjectMapper objectMapper = new ObjectMapper();
```
Comment (BerndBreier) [inline taskana-adapter-camunda-tasklistener/src/main/java/pro/taskana/adapter/camunda/tasklistener/TaskanaTaskListener.java]: Please don't make a new ObjectMapper for each call, but use a member (or static) variable for the ObjectMapper. Please note, that initialization must be complete before ObjectMapper is used for the first time.

File: taskana-adapter-camunda-system-connector/src/main/java/pro/taskana/adapter/systemconnector/camunda/api/impl/CamundaTaskRetriever.java [hunk 12]
```diff
@@ -38,7 +38,7 @@
 
     public List<ReferencedTask> retrieveNewStartedCamundaTasks(String camundaSystemTaskEventUrl) {
 
-        LOGGER.debug("entry to retrieveNewStartedCamundaTasks. createdAfter = {}");
+        LOGGER.debug("### entry to retrieveNewStartedCamundaTasks ###");
```
Comment (BerndBreier) [inline taskana-adapter-camunda-system-connector/src/main/java/pro/taskana/adapter/systemconnector/camunda/api/impl/CamundaTaskRetriever.java]: "###" strings in messages are fine during debugging. In Releases / production code they should be removed...

File: taskana-adapter-camunda-tasklistener/src/main/java/pro/taskana/adapter/camunda/tasklistener/TaskanaTaskListener.java (hunk 11 above)
Comment (gitgoodjhe) [reply taskana-adapter-camunda-tasklistener/src/main/java/pro/taskana/adapter/camunda/tasklistener/TaskanaTaskListener.java]: Done

File: taskana-adapter-camunda-system-connector/src/main/java/pro/taskana/adapter/systemconnector/camunda/api/impl/CamundaTaskRetriever.java (hunk 12 above)
Comment (gitgoodjhe) [reply taskana-adapter-camunda-system-connector/src/main/java/pro/taskana/adapter/systemconnector/camunda/api/impl/CamundaTaskRetriever.java]: Changed

**Pull request merged**